src/realm/sync/client_base.hpp

```diff
@@ -258,13 +258,16 @@ struct SessionErrorInfo : public ProtocolErrorInfo {
     {
     }
 
-    SessionErrorInfo(Status status, IsFatal is_fatal)
+    SessionErrorInfo(Status status, IsFatal is_fatal, void* error_user_code_callback = nullptr)
         : ProtocolErrorInfo(0, {}, is_fatal)
         , status(std::move(status))
+        , error_user_code_callback(error_user_code_callback) // propagate the error generated in the user code
```

Comment:
probably this is not the right approach, the errors in the user code callback have nothing to do with a generic sync session error, but this is the object we pass around to communicate possible issues, but it is solely used for communicating protocol errors and not user code errors.

Reply:
Ok, so the problem is thats sync errors (included the ones that are happening inside a user code callback) are bubbled up via `SessionErrorInfo`. 
So in case of user code exceptions, we don't know what is the type of the exception itself, it can be what ever in the SDK wrapping code.
So if we stick with this approach of storing the user code exception, we need this void ptr. However, we could simply store the exception string and let the SDK throw a new exception containing this string. 
This will simplify all this code.

Reply:
I don't have opinions about the `void*` and where it should be stored, but from usability perspective, we should aim to design the API in a way where we preserve the entire exception object rather than just the string. This is because the object will contain the exception type, the stacktrace, as well as inner exceptions.

Reply:
the ptr has been pushed back into the c api